### PR TITLE
Connection: use blog token to unlink users

### PR DESCRIPTION
**Pull request by sergeymitr**
#### Changes proposed in this Pull Request:
* Use blog token to unlink users from WPCOM, so if the user token is invalid, users can still get disconnected

Can only be merged after D89417-code.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
p1664981998624789-slack-C02NENKA4GN

#### Does this pull request change what data or activity we track or use?
No.

##### Secondary user
1. Connect Jetpack, create and connect a secondary user.
2. Create a secondary user, connect them to WPCOM. Open the "Blog Report" to confirm both users show up in the "Blog Users" section.
3. Go to Jetpack Dashboard, disconnect the secondary user, confirm they's been disconnected in the "Blog Report".
4. Reconnect the secondary user, confirm they show up in the "Blog Report"
5. Use the "Broken Token" tool to set an invalid user token for the secondary user.
6. Go back to Jetpack Dashboard, and disconnect the secondary user again.
7. Confirm the disconnect went well, you've seen the confirmation notice, and are now offered to connect the user.
8. Go to the "Blog Report" and confirm the user has been disconnected.

##### Connection owner
1. Connect Jetpack
2. Go to "Broken Token" and break the current user's token. Go to "Connection Errors" and create a "user token" verified error.
3. Go to Jetpack Dashboard and click "Restore Connection".
4. When you get redirected to the connection screen ("Unlock all the amazing features of Jetpack by connecting now"), open the "Broken Token" in another tab and confirm that the user's token and the "Primary User" are both empty.
5. Finish the connection flow, confirm it went well.

##### Jetpack CLI
1. Fully connect Jetpack
2. Run `wp jetpack disconnect user 1`, where `1` is the connection owner ID. Confirm that the user token and connection owner ID are **not** removed.
3. Run `wp jetpack disconnect user 1 --force`, confirm that the user token and connection owner ID are removed.

### CODE REVIEW

File: projects/packages/connection/src/class-manager.php
Comment (fgiannar) [inline projects/packages/connection/src/class-manager.php]: I think this should be under `if ( $is_disconnected_from_wpcom || $force_disconnect_locally ) {` as the rest of the user disconnect logic

**Review by fgiannar — CHANGES_REQUESTED** [review]
Many thanks for working on this, Sergey!

Left a small comment, most of the comments are in the WPCOM counterpart.
I think it makes more sense to resume testing this when the corresponding diff is merged.

File: projects/packages/connection/src/class-manager.php
Comment (sergeymitr) [reply projects/packages/connection/src/class-manager.php]: Good catch, thanks!

Comment (sergeymitr) [reply projects/packages/connection/src/class-manager.php]: Done https://github.com/Automattic/jetpack/pull/26705/commits/6c97416fd75336adf2215e0d88e6ce54eedba37e

**Review by fgiannar — DISMISSED** [review]
Let's :shipit: as soon as the WPCOM counterpart is shipped!

File: projects/packages/connection/src/class-manager.php
Comment (fgiannar) [inline projects/packages/connection/src/class-manager.php]: I dismissed my review as I realized we actually need to move this under `$is_disconnected_locally` check as `$this->get_tokens()->disconnect_user( $user_id, $can_overwrite_primary_user );` relies on the `master_user` option.

**Review by fgiannar — CHANGES_REQUESTED** [review]
Apart from the inline comments lets also make sure to fix the failing unit tests pls :)

**Comment by fgiannar** [comment]
@sergeymitr While reviewing this I also realized the logic for overwriting the primary user is rather wrong. We only check this for the local disconnection but always attempt to disconnect the user from WPCOM. I believe that if we don't allow for the primary user to be disconnected, then we shouldn't attempt to disconnect them from WPCOM, right?

**Comment by sergeymitr** [comment]
Hi @fgiannar.
Thanks for the review.
I moved the option removal to the `Token` class for it to run if the token actually gets removed: https://github.com/Automattic/jetpack/pull/26705/commits/8ef6248ff78ad75b21f1d796ff89c71bd8a3c746
Can you please take another look when you get a few minutes.

**Comment by fgiannar** [comment]
Mentioning here for visibility, we've discussed this with Sergey and he plans to implement the following as suggested by me above:

> @sergeymitr While reviewing this I also realized the logic for overwriting the primary user is rather wrong. We only check this for the local disconnection but always attempt to disconnect the user from WPCOM. I believe that if we don't allow for the primary user to be disconnected, then we shouldn't attempt to disconnect them from WPCOM, right?

Updating the status to reflect that.

**Comment by sergeymitr** [comment]
Hi @fgiannar.
I implemented some of the changes we discussed, please give it another look.
We can work on the `Token` refactoring and disconnecting users audit in separate PR's.

File: projects/packages/connection/src/class-tokens.php
Comment (fgiannar) [inline projects/packages/connection/src/class-tokens.php]: Do we still need this check here since we are deprecating `$can_overwrite_primary_user`?

Comment (sergeymitr) [reply projects/packages/connection/src/class-tokens.php]: @fgiannar since we're deprecating the argument, and not removing it, we need to let consumers know that they aren't supposed to use it now.
So I think adding the `_deprecated_argument()` notice is a good safeguard, even though I doubt there are any consumers of the function that still use the argument.

Comment (fgiannar) [reply projects/packages/connection/src/class-tokens.php]: 🤦  Sorry, I didn't mean the check for adding the notice. That's great! I meant this check: `if ( Jetpack_Options::get_option( 'master_user' ) === $user_id && ! $can_overwrite_primary_user ) {`

**Comment by leogermani** [comment]
FWIW, the "force disconnect" is used by the CLI command `jetpack disconnect user 1 --force`. Worth testing this as well

File: projects/packages/connection/src/class-tokens.php
Comment (sergeymitr) [reply projects/packages/connection/src/class-tokens.php]: As discussed in Slack, we'll keep the argument for now, and use the deprecation notice to inform potential consumers of the upcoming change.

**Comment by fgiannar** [comment]
> FWIW, the "force disconnect" is used by the CLI command `jetpack disconnect user 1 --force`. Worth testing this as well

Thanks, Leo! @sergeymitr  CLI uses [the Connection Manager to force disconnect](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/class.jetpack-cli.php#L274) a user so since that method still supports the argument we should be good.
In general, ideally the `Tokens` class should only be consumed by the Connection package.

**Review by fgiannar — DISMISSED** [review]
LGTM! Mostly looked at the code to approve (it's getting too late for me here) so in order to be on the safe side you might ask for another review (manual testing) before you merge!

**Comment by sergeymitr** [comment]
Thanks @leogermani, that helped discover an issue.

@fgiannar, leaving the `$can_overwrite_primary_user` didn't work out, as it prevented connection owners from being removed in certain cases.
So I followed your initial idea and removed the argument usage, only leaving the deprecation notice: https://github.com/Automattic/jetpack/pull/26705/commits/7176c6cfa960d579d43853e535dc20a835fff5a0

Please give it another look when you get a few minutes.

**Review by fgiannar — APPROVED** [review]
LGTM

**Pull request merged**
